src/librustc_mir/borrow_check/mod.rs

```diff
@@ -980,10 +980,27 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             // dropping a box does not touch data behind any
             // references it holds (which a Deep Write would touch).
             ty::TyAdt(def, _) if def.is_box() => {
+                // There are three main steps modelled here for
+                // dropping `a_box: Box<T`>`:
+                // 1. run the destructor for `*a_box`,
+                // 2. deallocate the backing storage of `*a_box`,
+                // 3. the StorageDead of `a_box` itself
+                //
+                // When/if we add a `&own T` type, step 1 would be
+                // like running the destructor of the `&own T` (and
+                // the owner of backing storage referenced by the
+                // `&own T` would be responsible for steps 2 and 3).
```

Comment:
At the lowest (unsafe?) level, yes, step 2 would *technically* be in `Drop for Box<T>`.
*However*, from the point of view of the code using `DerefOwn` + `Drop`, step 2 would be *part of* dropping the `&own T`.

That is, dropping `&own T` invalidates it as backing storage for the code using it, so there would potentially be a gap in between dropping the `&own T` you got from `DerefOwn` and calling `free` (via `Drop for Box<T>`), in which *you can't use the allocation* even though it still exists.

This is *quite distinct* from *moving out* of the `&own T`, which would allow moving back in, or assigning to the `T` inside, which would drop the `T` *but not invalidate* the `&own T`.

Reply:
Hmm so you’re interpreting step 2 as solely invalidating the backing storage, while I was interpreting it as both invalidating the backing storage *and* deallocating the backing storage

Do I have that right? Where does the deallocation happen in your mental model, as part of step 3?

Reply:
Or, rereading my comment in the code, maybe I was interpreting step 2 as solely deallocating the backing storage, and I didn’t have the invalidation in any explicit step.

While you are pointing out that the invalidation happens, in the code here, as part of step 2

Reply:
> While you are pointing out that the invalidation happens, in the code here, as part of step 2

Yes, in terms of borrowck, you *can only* observe that "backing storage" invalidation, since the deallocation is a FFI call with a raw pointer, it's not something borrowck can understand directly.

Reply:
Talking to @RalfJung, I reached a point where I'm wondering whether `Drop(x: T)`  being a noop when `!needs_drop(T)` is a bad idea, since it makes some sense that it'd have an effect similar to `Operand::Move(x)` (i.e. invalidation of the source value).

I thought something like `StorageDead(*a_box)` was needed, but `Move(*a_box)` should do and `Drop(*x)` should *maybe* imply it as well.
For the `&own T` model, that means that `Drop(x: &own T)` only needs to invalidate *the value at* `*x`, *not* the underlying memory. However, that memory is no longer reachable, as `x` had the unique path to it (which the `Drop(x)` should have invalidated), and invalidating the value at `*x` should prevent any other borrows from pointing to it.

In less words, we only need `StorageLive` / `StorageDead` for locals *specifically because* they're not behind indirection we can create/destroy in ways that would directly enable/disable access to the value.

OTOH we don't generate `Drop` at all sometimes, as an optimization, so we have to rely on `StorageDead` instead for invalidating at least *some* of the locals.